### ISSUE TITLE
Singlet sequences

### ISSUE THREAD

**Issue by deprekate** [voutcn/megahit#56]
Works great so far, but I don't see a file (or a command line option to produce a file) with the singlets, or any annotation file of which reads went into which contigs (which would allow me to get the singlet reads).

**Comment by voutcn** [comment]
We don't provide such an option. You could aligned reads back to the contigs with short-read aligners say Bowtie2.

**Comment by deprekate** [comment]
It might be good to tag this as a bug/feature request?   One of the standard pipelines for metagenome analysis is to assemble, then dump the singlet reads in with the contigs, and then do downstream analysis.

As an example I just grabbed the newest uploaded WGS metagenome on MGRAST (4640988.3)
I assembled using MIRA to find out approximatley how many reads do not assemble into contigs.  Out of 519,098 input reads, less than half assembled into contigs.   It seems unwise to discard half of your data.
```
$ grep -c "^>" mgm4640988.3.050.upload.fna 
519098

$ wc -l Test_info_contigreadlist.txt 
254040

$ wc -l Test_info_debrislist.txt 
265058
```

And since which reads went into which contigs is also not reported, there is no way to calculate coverage.  This means you cannot do abundance calculations on the data (or use this data on MGRAST).

**Comment by voutcn** [comment]
It is not a bug. Knowing little about MIRA, but I think most de Bruijn graph based assemblers would not report which reads goes to which contigs, because reads have been already chopped into kmers and kmers are de-duplicated.

One could easily aligned the reads back to the contigs and output the unmapped reads for subsequent analysis. As I know Bowtie2 provides some options to output unmapped reads. I would probably write a tutorial on this in the near future.

**Comment by epruesse** [comment]
bbmap can do it like so:

bbmap in=read_1.fq in2=read_2.fq ref=final.contigs.fa nodisk outu=unmapped_reads_1.fq outu2=unmapped_reads_2.fq

You might want to change the stringency of the mapping though.

On Jul 29, 2015, at 12:00 AM, Dinghua Li <notifications@github.com<mailto:notifications@github.com>> wrote:


It is not a bug. Knowing little about MIRA, but I think most de Bruijn graph based assemblers would not report which reads goes to which contigs, because reads have been already chopped into kmers and kmers are de-duplicated.

One could easily aligned the reads back to the contigs and output the unmapped reads for subsequent analysis. As I know Bowtie2 provides some options to output unmapped reads. I would probably write a tutorial on this in the near future.

**Comment by voutcn** [comment]
Thanks @epruesse. I dicided to write a simple tutorial based on BBMap.

**Comment by deprekate** [comment]
Cool, I got bbmap to work.  Out of my test assembly of 30,000 sequences, MEGAHIT/bbmap had 1,805 singlets.  I used CAP3 (as it is very simplistic, but super slow) to get a basis of comparison, and CAP3 assembly had 1,580 singlets, which is about the same as MEGAHIT/bbmap.  So only tweaking the stringency to match default settings of MEGAHIT is needed.  

Where is the writeup tutorial?   I am now trying to figure out how to get the coverage of each contig from bbmap, since I believe that is not something estimated by MEGAHIT.   I tried different bbmap "Histogram and statistics output parameters:" but all gave weird results (lots of zero coverage, when techincally a contig should have had at least two reads go into it)
